// @tryghost/ghost/build/nestjs/modules/app.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppModule = void 0;
const core_1 = require("@nestjs/core");
const admin_api_module_1 = require("./admin-api.module");
const not_found_fallthrough_filter_1 = require("../filters/not-found-fallthrough.filter");
const example_listener_1 = require("../../listeners/example.listener");
const global_exception_filter_1 = require("../filters/global-exception.filter");
class AppModuleClass {
}
exports.AppModule = {
    global: true,
    module: AppModuleClass,
    imports: [
        core_1.RouterModule.register([
            {
                path: 'ghost/api/admin',
                module: admin_api_module_1.AdminAPIModule
            }
        ]),
        admin_api_module_1.AdminAPIModule
    ],
    exports: [],
    controllers: [],
    providers: [
        example_listener_1.ExampleListener,
        {
            provide: core_1.APP_FILTER,
            useClass: global_exception_filter_1.GlobalExceptionFilter
        }, {
            provide: core_1.APP_FILTER,
            useClass: not_found_fallthrough_filter_1.NotFoundFallthroughExceptionFilter
        }
    ]
};
